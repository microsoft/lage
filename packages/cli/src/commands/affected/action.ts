import createLogger from "@lage-run/logger";
import { getPackageInfos, getWorkspaceRoot, type PackageInfos } from "workspace-tools";
import { getConfig } from "../../config/getConfig.js";
import { getFilteredPackages } from "../../filter/getFilteredPackages.js";
import type { FilterOptions } from "../../types/FilterOptions.js";

interface AffectedOptions extends FilterOptions {
  outputFormat?: "json" | "graph" | "default";
}

export async function affectedAction(options: AffectedOptions) {
  const { dependencies, dependents, since, scope, ignore, outputFormat } = options;

  const cwd = process.cwd();
  const config = await getConfig(cwd);
  const logger = createLogger();

  const root = getWorkspaceRoot(cwd)!;
  const packageInfos = getPackageInfos(root);

  const packages = getFilteredPackages({
    root,
    logger,
    packageInfos,
    includeDependencies: dependencies,
    includeDependents: dependents,
    since,
    scope,
    repoWideChanges: config.repoWideChanges,
    sinceIgnoreGlobs: ignore,
  });

  let output = "";
  switch (outputFormat) {
    case "graph":
      output = renderGraph({ packages, packageInfos });
      break;

    case "json":
      output = renderJson({ packages, packageInfos });
      break;

    default:
      output = renderDefault({ packages });
      break;
  }

  // eslint-disable-next-line no-console
  console.log(output);
}

function renderDefault(props: { packages: string[] }) {
  const { packages } = props;
  return `
All Affected Packages
---------------------

${packages.join("\n")}
`;
}

function renderJson(props: { packages: string[]; packageInfos: PackageInfos }) {
  const graph = generatePackageGraph(props);
  return JSON.stringify(graph);
}

function renderGraph(props: { packages: string[]; packageInfos: PackageInfos }) {
  const graph = generatePackageGraph(props);
  const { packages } = graph;

  const adjacencies: [string, string][] = [];

  for (const [pkg, info] of Object.entries(packages)) {
    for (const dep of info.dependencies) {
      adjacencies.push([pkg, dep]);
    }
  }

  return `
digraph affected {
${adjacencies.map((entry) => `  "${entry[0]}" -> "${entry[1]}"`).join("\n")}
}  
`;
}

function generatePackageGraph(props: { packages: string[]; packageInfos: PackageInfos }) {
  const { packages, packageInfos } = props;
  const packageGraph = packages.reduce<{ [pkg: string]: { dependencies: string[]; dependents: string[] } }>((accum, pkg) => {
    const dependencies = Object.keys(packageInfos[pkg].dependencies ?? {}).filter((dep) => packages.includes(dep));
    const dependents = Object.keys(packageInfos[pkg].dependents ?? {}).filter((dep) => packages.includes(dep));

    accum[pkg] = {
      dependencies,
      dependents,
    };

    return accum;
  }, {});

  return { packages: packageGraph, count: packages.length };
}
