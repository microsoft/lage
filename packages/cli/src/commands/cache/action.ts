import { clearCache } from "./clearCache.js";
import type { Command } from "commander";
import { getConfig } from "@lage-run/config";
import { pruneCache } from "./pruneCache.js";
import createLogger from "@lage-run/logger";
import type { ReporterInitOptions } from "../../types/ReporterInitOptions.js";

interface CacheOptions extends ReporterInitOptions {
  prune?: number;
  clear?: boolean;
}

export async function cacheAction(options: CacheOptions, command: Command) {
  const cwd = process.cwd();
  const config = await getConfig(cwd);
  const logger = createLogger();

  if (options.clear) {
    return await clearCache({
      cwd: process.cwd(),
      internalCacheFolder: config.cacheOptions.internalCacheFolder,
      logger,
      concurrency: options.concurrency,
    });
  } else if (options.prune) {
    return await pruneCache({
      pruneDays: options.prune,
      cwd: process.cwd(),
      internalCacheFolder: config.cacheOptions.internalCacheFolder,
      logger,
      concurrency: options.concurrency,
    });
  }

  command.help();
}
