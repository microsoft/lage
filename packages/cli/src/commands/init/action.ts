/* eslint-disable no-console -- logger doesn't work in this context */
import { readConfigFile } from "@lage-run/config";
import fs from "fs";
import path from "path";
import execa from "execa";

type WorkspaceManager = "rush" | "pnpm" | "yarn" | "npm";

export async function initAction() {
  const cwd = process.cwd();

  const config = await readConfigFile(cwd);
  if (config) {
    console.error("lage is already initialized in this workspace");
    process.exitCode = 1;
    return;
  }

  console.info("Installing lage and creating a default configuration file");

  let workspaceManager: WorkspaceManager = "yarn";

  try {
    workspaceManager = whichWorkspaceManager(cwd);
  } catch (e) {
    console.error(
      "lage requires you to be using a workspace - make sure you are using yarn workspaces, npm workspaces, pnpm workspaces, or rush"
    );
    process.exitCode = 1;
    return;
  }

  const pipeline = {
    build: ["^build"],
    test: ["build"],
    lint: [],
  };

  const lageConfig = {
    pipeline,
    npmClient: workspaceManager === "yarn" ? "yarn" : "npm",
  };

  const lageConfigFile = path.join(cwd, "lage.config.js");
  fs.writeFileSync(lageConfigFile, "module.exports = " + JSON.stringify(lageConfig, null, 2) + ";");

  installLage(cwd, workspaceManager, pipeline);

  console.info(`Lage is initialized! You can now run: ${getBuildCommand(workspaceManager)}`);
}

function getBuildCommand(workspaceManager: WorkspaceManager) {
  switch (workspaceManager) {
    case "yarn":
      return "yarn lage build";

    case "pnpm":
      return "pnpm run lage build";

    case "rush":
    case "npm":
      return "npm run lage build";
  }
}

function whichWorkspaceManager(cwd: string) {
  const packageJson = readPackageJson(cwd);

  if (fs.existsSync(path.join(cwd, "rush.json"))) {
    return "rush";
  }

  if (fs.existsSync(path.join(cwd, "yarn.lock")) && packageJson.workspaces) {
    return "yarn";
  }

  if (fs.existsSync(path.join(cwd, "pnpm-workspace.yaml"))) {
    return "pnpm";
  }

  if (fs.existsSync(path.join(cwd, "package-lock.json")) && packageJson.workspaces) {
    return "npm";
  }

  throw new Error("not a workspace");
}

async function installLage(cwd: string, workspaceManager: WorkspaceManager, pipeline: Record<string, string[]>) {
  const lageVersion = getLageVersion();
  const packageJson = readPackageJson(cwd);
  packageJson.scripts ??= {};
  for (const script of Object.keys(pipeline)) {
    packageJson.scripts[script] = `lage ${script}`;
  }

  if (workspaceManager === "rush") {
    packageJson.scripts.lage = `node common/scripts/install-run.js lage@${lageVersion} lage`;
    writePackageJson(cwd, packageJson);
  } else {
    packageJson.scripts.lage = "lage";
    packageJson.devDependencies ??= {};
    packageJson.devDependencies.lage = lageVersion;
    writePackageJson(cwd, packageJson);

    await execa(workspaceManager, ["install"], { stdio: "inherit" });
  }
}

function getLageVersion() {
  const lagePackageJsonFile = require.resolve("../../package.json", {
    paths: [__dirname],
  });
  const lagePackageJson = JSON.parse(fs.readFileSync(lagePackageJsonFile, "utf-8"));
  return lagePackageJson.version;
}

function writePackageJson(cwd: string, packageJson: any) {
  const packageJsonFile = path.join(cwd, "package.json");
  fs.writeFileSync(packageJsonFile, JSON.stringify(packageJson, null, 2));
}

function readPackageJson(cwd: string): {
  scripts?: Record<string, string>;
  devDependencies?: Record<string, string>;
  workspaces?: any;
} {
  const packageJsonFile = path.join(cwd, "package.json");
  return JSON.parse(fs.readFileSync(packageJsonFile, "utf-8"));
}
