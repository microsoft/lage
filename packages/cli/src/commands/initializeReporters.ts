import { createReporter } from "./createReporter.js";
import type { Logger } from "@lage-run/logger";
import type { ReporterInitOptions } from "../types/ReporterInitOptions.js";

export function initializeReporters(logger: Logger, options: ReporterInitOptions) {
  const { reporter } = options;

  // filter out falsy values (e.g. undefined) from the reporter array
  const reporterOptions = (Array.isArray(reporter) ? reporter : [reporter]).filter(Boolean);

  if (reporterOptions.length === 0) {
    // "default" is just a dummy name to trigger the default case in createReporter
    reporterOptions.push("default");
  }

  // add profile reporter if --profile is passed
  if (options.profile) {
    reporterOptions.push("profile");
  }

  for (const reporterName of reporterOptions) {
    const reporterInstance = createReporter(reporterName, options);
    logger.addReporter(reporterInstance);
  }

  return logger.reporters;
}
