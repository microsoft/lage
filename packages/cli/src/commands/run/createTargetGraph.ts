import type { Logger } from "@lage-run/logger";
import { TargetGraphBuilder } from "@lage-run/target-graph";
import type { PackageInfos } from "workspace-tools";
import { getFilteredPackages } from "../../filter/getFilteredPackages.js";
import type { PipelineDefinition } from "../../types/PipelineDefinition.js";

interface CreateTargetGraphOptions {
  logger: Logger;
  root: string;
  dependencies: boolean;
  dependents: boolean;
  since: string;
  scope: string[];
  ignore: string[];
  repoWideChanges: string[];
  pipeline: PipelineDefinition;
  outputs: string[];
  tasks: string[];
  packageInfos: PackageInfos;
}

export function createTargetGraph(options: CreateTargetGraphOptions) {
  const { logger, root, dependencies, dependents, since, scope, repoWideChanges, ignore, pipeline, outputs, tasks, packageInfos } = options;

  const builder = new TargetGraphBuilder(root, packageInfos);

  const packages = getFilteredPackages({
    root,
    logger,
    packageInfos,
    includeDependencies: dependencies,
    includeDependents: dependents,
    since,
    scope,
    repoWideChanges,
    sinceIgnoreGlobs: ignore,
  });

  for (const [id, definition] of Object.entries(pipeline)) {
    if (Array.isArray(definition)) {
      builder.addTargetConfig(id, {
        cache: true,
        dependsOn: definition,
        options: {},
        outputs,
      });
    } else {
      builder.addTargetConfig(id, definition);
    }
  }

  return builder.buildTargetGraph(tasks, packages);
}
