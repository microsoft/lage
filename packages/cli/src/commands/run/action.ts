import { BackfillCacheProvider, RemoteFallbackCacheProvider, TargetHasher } from "@lage-run/cache";
import { Command } from "commander";
import { createProfileReporter } from "./createProfileReporter";
import { findNpmClient } from "@lage-run/find-npm-client";
import { getConfig } from "../../config/getConfig";
import { getFilteredPackages } from "../../filter/getFilteredPackages";
import { getMaxWorkersPerTask } from "../../config/getMaxWorkersPerTask";
import { getPackageInfos, getWorkspaceRoot } from "workspace-tools";
import { initializeReporters } from "@lage-run/reporters";
import { isRunningFromCI } from "../isRunningFromCI";
import { SimpleScheduler } from "@lage-run/scheduler";
import { TargetGraphBuilder } from "@lage-run/target-graph";
import createLogger from "@lage-run/logger";
import type { ReporterInitOptions } from "@lage-run/reporters";

function filterArgsForTasks(args: string[]) {
  const optionsPosition = args.findIndex((arg) => arg.startsWith("-"));
  return {
    tasks: args.slice(0, optionsPosition === -1 ? undefined : optionsPosition),
    taskArgs: optionsPosition === -1 ? [] : args.slice(optionsPosition),
  };
}

interface RunOptions extends ReporterInitOptions {
  concurrency: number;
  profile: string | boolean | undefined;
  dependencies: boolean;
  dependents: boolean;
  since: string;
  scope: string[];
  skipLocalCache: boolean;
  continue: boolean;
  cache: boolean;
  resetCache: boolean;
  nodeArg: string;
  ignore: string[];
}

export async function runAction(options: RunOptions, command: Command) {
  const cwd = process.cwd();
  const config = await getConfig(cwd);

  // Configure logger
  const logger = createLogger();

  initializeReporters(logger, options);

  if (options.profile !== undefined) {
    const reporter = createProfileReporter(options);
    logger.addReporter(reporter);
  }

  // Build Target Graph
  const root = getWorkspaceRoot(process.cwd())!;
  const packageInfos = getPackageInfos(root);

  const builder = new TargetGraphBuilder(root, packageInfos);

  const { tasks, taskArgs } = filterArgsForTasks(command.args);

  const packages = getFilteredPackages({
    root,
    logger,
    packageInfos,
    includeDependencies: options.dependencies,
    includeDependents: options.dependents,
    since: options.since,
    scope: options.scope,
    repoWideChanges: config.repoWideChanges,
    sinceIgnoreGlobs: options.ignore.concat(config.ignore),
  });

  for (const [id, definition] of Object.entries(config.pipeline)) {
    if (Array.isArray(definition)) {
      builder.addTargetConfig(id, {
        cache: true,
        dependsOn: definition,
        options: {},
        outputs: config.cacheOptions.outputGlob,
      });
    } else {
      builder.addTargetConfig(id, definition);
    }
  }

  const targetGraph = builder.buildTargetGraph(tasks, packages);

  const hasRemoteCacheConfig =
    !!config.cacheOptions?.cacheStorageConfig || !!process.env.BACKFILL_CACHE_PROVIDER || !!process.env.BACKFILL_CACHE_PROVIDER_OPTIONS;

  // Create Cache Provider
  const cacheProvider = new RemoteFallbackCacheProvider({
    root,
    logger,
    localCacheProvider:
      options.skipLocalCache === true
        ? undefined
        : new BackfillCacheProvider({
            logger,
            root,
            cacheOptions: {
              outputGlob: config.cacheOptions.outputGlob,
              ...(config.cacheOptions.internalCacheFolder && { internalCacheFolder: config.cacheOptions.internalCacheFolder }),
            },
          }),
    remoteCacheProvider: hasRemoteCacheConfig ? new BackfillCacheProvider({ logger, root, cacheOptions: config.cacheOptions }) : undefined,
    writeRemoteCache:
      config.cacheOptions?.writeRemoteCache === true || String(process.env.LAGE_WRITE_CACHE).toLowerCase() === "true" || isRunningFromCI,
  });

  const hasher = new TargetHasher({
    root,
    environmentGlob: config.cacheOptions.environmentGlob,
    cacheKey: config.cacheOptions.cacheKey,
  });

  const scheduler = new SimpleScheduler({
    logger,
    concurrency: options.concurrency,
    cacheProvider,
    hasher,
    continueOnError: options.continue,
    shouldCache: options.cache,
    shouldResetCache: options.resetCache,
    maxWorkersPerTask: getMaxWorkersPerTask(config.pipeline ?? {}),
    runners: {
      npmScript: {
        script: require.resolve("./runners/NpmScriptRunner"),
        options: {
          nodeArg: options.nodeArg,
          taskArgs,
          npmCmd: findNpmClient(config.npmClient),
        },
      },
      worker: {
        script: require.resolve("./runners/WorkerRunner"),
        options: {},
      },
      ...config.runners,
    },
  });

  const summary = await scheduler.run(root, targetGraph);

  if (summary.results !== "success") {
    process.exitCode = 1;
  }

  for (const reporter of logger.reporters) {
    reporter.summarize(summary);
  }
}
