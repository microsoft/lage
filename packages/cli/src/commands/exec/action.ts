import type { Command } from "commander";
import createLogger from "@lage-run/logger";
import type { ReporterInitOptions } from "../../types/ReporterInitOptions.js";
import { initializeReporters } from "../initializeReporters.js";
import { executeInProcess } from "./executeInProcess.js";
import { executeRemotely } from "./executeRemotely.js";

interface ExecOptions extends ReporterInitOptions {
  cwd?: string;
  server?: boolean | string;
  timeout?: number;
  nodeArg?: string;
  tasks?: string[];
}

interface ExecRemoteOptions extends ExecOptions {
  tasks: string[];
}

export async function execAction(options: ExecOptions, command: Command) {
  const logger = createLogger();
  options.cwd = options.cwd ?? process.cwd();
  options.logLevel = options.logLevel ?? "info";
  options.reporter = options.reporter ?? "json";
  initializeReporters(logger, options);

  const { server } = options;
  if (server) {
    logger.info("Running in server mode");

    if (typeof options.tasks === "undefined") {
      throw new Error("No tasks specified, this is required for when running in server mode");
    }

    await executeRemotely(options as ExecRemoteOptions, command);
  } else {
    await executeInProcess({ logger, args: command.args, cwd: options.cwd, nodeArg: options.nodeArg });
  }
}
