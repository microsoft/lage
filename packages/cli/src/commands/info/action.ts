import type { Command } from "commander";
import { createTargetGraph } from "../run/createTargetGraph.js";
import { filterArgsForTasks } from "../run/filterArgsForTasks.js";
import { getConfig } from "@lage-run/config";
import { getPackageInfosAsync, getWorkspaceRoot } from "workspace-tools";
import { getFilteredPackages } from "../../filter/getFilteredPackages.js";
import createLogger from "@lage-run/logger";
import path from "path";

import type { ReporterInitOptions } from "../../types/ReporterInitOptions.js";
import { getStartTargetId } from "@lage-run/target-graph";
import { initializeReporters } from "../initializeReporters.js";

interface RunOptions extends ReporterInitOptions {
  dependencies: boolean;
  dependents: boolean;
  since: string;
  scope: string[];
  to: string[];
  cache: boolean;
  nodeArg: string;
  ignore: string[];
}

interface PackageTask {
  id: string;
  command: string[];
  dependencies: string[];
  workingDirectory: string;
  package: string;
  task: string;
}

/**
 * (UNSTABLE) The info command displays information about a target graph in a workspace.
 * The generated output can be read and used by other task runners, such as BuildXL.
 *
 * Expected format:
 * [
 *   {
 *       "id": "bar##build",
 *       "package": "bar",
 *       "task": "build",
 *       "command": "npm run build --blah",
 *       "workingDirectory": "packages/bar",
 *       "dependencies": []
 *   },
 *   {
 *       "id": "foo##build",
 *       "package": "foo",
 *       "task": "build",
 *       "command": "npm run build --blah",
 *       "workingDirectory": "packages/foo",
 *       "dependencies": [
 *           "bar##build"
 *       ]
 *   },
 *   {
 *       "id": "foo##test",
 *       "package": "foo",
 *       "task": "test",
 *       "command": "npm run test --blah",
 *       "workingDirectory": "packages/foo",
 *       "dependencies": [
 *           "foo##build"
 *       ]
 *   },
 *   ...
 * ]
 */
export async function infoAction(options: RunOptions, command: Command) {
  const cwd = process.cwd();
  const config = await getConfig(cwd);
  const logger = createLogger();
  options.logLevel = options.logLevel ?? "info";
  options.reporter = options.reporter ?? "json";
  initializeReporters(logger, options);
  const root = getWorkspaceRoot(cwd)!;

  const packageInfos = await getPackageInfosAsync(root);
  const targetGraph = prepareAndCreateTargetGraph(config, logger, root, options, packageInfos, command);
  const scope = prepareAndGetFilteredPackages(config, logger, root, options, packageInfos);
  const packageTasks = processTargets(targetGraph.targets, packageInfos, config);

  logger.info("info", {
    command: command.args,
    scope,
    packageTasks: [...packageTasks.values()].flat(),
  });
}

function prepareAndCreateTargetGraph(config, logger, root, options, packageInfos, command) {
  const { tasks } = filterArgsForTasks(command.args);
  const targetGraph = createTargetGraph({
    logger,
    root,
    dependencies: options.dependencies,
    dependents: options.dependents && !options.to, // --to is a short hand for --scope + --no-dependents
    ignore: options.ignore.concat(config.ignore),
    pipeline: config.pipeline,
    repoWideChanges: config.repoWideChanges,
    scope: (options.scope ?? []).concat(options.to ?? []), // --to is a short hand for --scope + --no-dependents
    since: options.since,
    outputs: config.cacheOptions.outputGlob,
    tasks,
    packageInfos,
  });

  return targetGraph;
}

function prepareAndGetFilteredPackages(config, logger, root, options, packageInfos) {
  const scope = getFilteredPackages({
    root,
    packageInfos,
    logger,
    includeDependencies: options.dependencies,
    includeDependents: options.dependents && !options.to, // --to is a short hand for --scope + --no-dependents
    since: options.since,
    scope: (options.scope ?? []).concat(options.to ?? []), // --to is a short hand for --scope + --no-dependents
    repoWideChanges: config.repoWideChanges,
    sinceIgnoreGlobs: options.ignore.concat(config.ignore),
  });

  return scope;
}

function processTargets(targets, packageInfos, config) {
  const packageTasks = new Map<string, PackageTask[]>(); // Initialize the map with the correct type
  const dependenciesCache = new Map<string, string[]>();

  for (const target of targets.values()) {
    if (shouldSkipTarget(target, packageInfos)) {
      continue;
    }

    const packageTask = generatePackageTask(target, targets, packageInfos, dependenciesCache, config);
    if (packageTask) {
      // Check if the packageTask is defined before accessing its properties
      const packageName = packageTask.package;
      if (!packageTasks.has(packageName)) {
        packageTasks.set(packageName, []);
      }
      packageTasks.get(packageName)!.push(packageTask); // Use the non-null assertion operator to avoid type errors
    }
  }

  return packageTasks;
}

function isTargetNoop(target, packageInfos) {
  return !packageInfos[target.packageName]?.scripts?.[target.task];
}

function shouldSkipTarget(target, packageInfos) {
  return target.id === getStartTargetId() || isTargetNoop(target, packageInfos);
}

function generatePackageTask(target, targets, packageInfos, dependenciesCache, config): PackageTask {
  const command = generateCommand(target, config);
  const workingDirectory = getWorkingDirectory(target);

  const dependenciesSet = resolveDependencies(target.dependencies, targets, packageInfos, dependenciesCache);

  const packageTask: PackageTask = {
    id: target.id,
    command,
    dependencies: [...dependenciesSet],
    workingDirectory,
    package: target.packageName,
    task: target.task,
  };

  return packageTask;
}

function resolveDependencies(dependencies: string[], targets, packageInfos, dependenciesCache: Map<string, Set<string>>) {
  const result: Set<string> = new Set();

  for (const dependency of dependencies) {
    if (dependency === getStartTargetId()) {
      continue;
    }

    if (!dependenciesCache.has(dependency)) {
      const dependencyTarget = targets.get(dependency);
      if (isTargetNoop(dependencyTarget, packageInfos)) {
        dependenciesCache.set(dependency, resolveDependencies(dependencyTarget.dependencies, targets, packageInfos, dependenciesCache));
      } else {
        dependenciesCache.set(dependency, new Set([dependency]));
      }
    }

    dependenciesCache.get(dependency)?.forEach((dependency: string) => result.add(dependency));
  }

  return result;
}

function generateCommand(target, config) {
  const npmClient = config.npmClient ?? "npm";

  const command = [npmClient, ...getNpmArgs(target.task, target.taskArgs)];
  return command;
}

function getWorkingDirectory(target) {
  const cwd = process.cwd();
  const workingDirectory = path.relative(getWorkspaceRoot(cwd) ?? "", target.cwd).replace(/\\/g, "/");
  return workingDirectory;
}

function getNpmArgs(task: string, taskTargs: string[]) {
  const extraArgs = taskTargs != undefined && taskTargs.length > 0 ? ["--", ...taskTargs] : [];
  return ["run", task, ...extraArgs];
}
