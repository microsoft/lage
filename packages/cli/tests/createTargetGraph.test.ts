import { Logger } from "@lage-run/logger";
import { getConfig, PipelineDefinition } from "@lage-run/config";
import createLogger from "@lage-run/logger";
import { initializeReporters } from "../src/commands/initializeReporters.js";
import { generatePackageTask, type InfoActionOptions } from "../src/commands/info/action.js";
import { PackageInfo, PackageInfos } from "workspace-tools";
import { createTargetGraph } from "../src/commands/run/createTargetGraph.js";
import { getFilteredPackages } from "../src/filter/getFilteredPackages.js";
import { getBinPaths } from "../src/getBinPaths.js";

describe("createTargetGraph", () => {
  const logger = new Logger();

  it("Basic graph, seperate nodes", async () => {
    const packageInfos: PackageInfos = {
      foo1: stubPackage({ name: "foo1", deps: ["foo2"] }),
      foo2: stubPackage({ name: "foo2" }),
    };

    const pipeline: PipelineDefinition = {
      build: ["^build"],
      test: ["build"],
    };

    const result = await createAndPrintPackageTasks(["build"], packageInfos, pipeline, ["id", "dependencies"]);
    expect(result).toMatchSnapshot();
  });

  it("Basic graph, spanning tasks", async () => {
    const packageInfos: PackageInfos = {
      foo1: stubPackage({ name: "foo1", deps: ["foo2"] }),
      foo2: stubPackage({ name: "foo2" }),
    };

    const pipeline: PipelineDefinition = {
      build: ["^build"],
      test: ["build"],
    };

    const result = await createAndPrintPackageTasks(["build", "test"], packageInfos, pipeline, ["id", "dependencies"]);
    expect(result).toMatchSnapshot();
  });

  it("PackageJsonOverrideForTargetDeps", async () => {
    const packageInfos: PackageInfos = {
      foo1: stubPackage({ name: "foo1", deps: ["foo2"] }),
      foo2: stubPackage({ name: "foo2", fields: { lage: { test: ["build"] } } }),
    };

    const pipeline: PipelineDefinition = {
      build: ["^build"],
      test: [],
    };

    const result = await createAndPrintPackageTasks(["build", "test"], packageInfos, pipeline, ["id", "dependencies"]);
    expect(result).toMatchSnapshot();
  });
});

async function createAndPrintPackageTasks(
  tasks: string[],
  packageInfos: PackageInfos,
  pipeline: PipelineDefinition,
  fields: string[]
): Promise<string> {
  const packageTasks = await createPackageTasks(tasks, packageInfos, pipeline);
  const expected = filterObjects(packageTasks, ["id", "dependencies"]);
  return JSON.stringify(expected, null, 2);
}

function filterObjects<T extends object>(objects: T[], fields: string[]): Partial<T>[] {
  return objects.map((obj) => filterFields(obj, fields));
}

function filterFields<T extends object>(obj: T, fields: string[]): Partial<T> {
  return <Partial<T>>Object.fromEntries(Object.entries(obj).filter(([key]) => fields.includes(key)));
}

async function createPackageTasks(tasks: string[], packageInfos: PackageInfos, pipeline: PipelineDefinition): Promise<PackageTask[]> {
  const root = process.cwd();
  const config = await getConfig(root);
  config.pipeline = pipeline;
  const logger = createLogger();
  const options: InfoActionOptions = {
    logLevel: "info",
    reporter: "json",
    dependencies: false,
    dependents: false,
    since: <string>(<any>undefined),
    scope: [],
    to: [],
    cache: false,
    nodeArg: "",
    ignore: [],
    server: "",
    progress: false,
    verbose: false,
    grouped: false,
    concurrency: 1,
  };

  initializeReporters(logger, options);

  const targetGraph = await createTargetGraph({
    logger,
    root,
    dependencies: options.dependencies,
    dependents: options.dependents && !options.to, // --to is a short hand for --scope + --no-dependents
    ignore: options.ignore.concat(config.ignore),
    pipeline: config.pipeline,
    repoWideChanges: config.repoWideChanges,
    scope: (options.scope ?? []).concat(options.to ?? []), // --to is a short hand for --scope + --no-dependents
    since: options.since,
    outputs: config.cacheOptions.outputGlob,
    tasks,
    packageInfos,
  });

  const scope = getFilteredPackages({
    root,
    packageInfos,
    logger,
    includeDependencies: options.dependencies,
    includeDependents: options.dependents && !options.to, // --to is a short hand for --scope + --no-dependents
    since: options.since,
    scope: (options.scope ?? []).concat(options.to ?? []), // --to is a short hand for --scope + --no-dependents
    repoWideChanges: config.repoWideChanges,
    sinceIgnoreGlobs: options.ignore.concat(config.ignore),
  });

  const binPaths = getBinPaths();
  const targets = [...targetGraph.targets.values()];
  const packageTasks = targets.map((target) => generatePackageTask(target, [], config, options, binPaths, packageInfos, tasks));

  return packageTasks;
}

function stubPackage(args: { name: string; deps?: string[]; scripts?: string[]; fields?: object }): PackageInfo {
  return {
    name: args.name,
    packageJsonPath: `packages/${args.name}`,
    version: "1.0",
    dependencies: (args.deps || []).reduce((depMap, dep) => ({ ...depMap, [dep]: "*" }), {}),
    devDependencies: {},
    scripts: (args.scripts || []).reduce((scriptMap, script) => ({ ...scriptMap, [script]: `echo ${script}` }), {}),
    ...(args.fields || {}),
  } as PackageInfo;
}
