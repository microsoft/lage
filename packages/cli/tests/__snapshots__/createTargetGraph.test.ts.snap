// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTargetGraph Basic graph, seperate nodes 1`] = `
"[
  {
    "id": "__start",
    "dependencies": []
  },
  {
    "id": "foo1#build",
    "dependencies": [
      "__start",
      "foo2#build"
    ]
  },
  {
    "id": "foo2#build",
    "dependencies": [
      "__start"
    ]
  }
]"
`;

exports[`createTargetGraph Basic graph, spanning tasks 1`] = `
"[
  {
    "id": "__start",
    "dependencies": []
  },
  {
    "id": "foo1#build",
    "dependencies": [
      "__start",
      "foo2#build"
    ]
  },
  {
    "id": "foo2#build",
    "dependencies": [
      "__start"
    ]
  },
  {
    "id": "foo1#test",
    "dependencies": [
      "__start",
      "foo1#build"
    ]
  },
  {
    "id": "foo2#test",
    "dependencies": [
      "__start",
      "foo2#build"
    ]
  }
]"
`;

exports[`createTargetGraph PackageJsonOverrideForTargetDeps 1`] = `
"[
  {
    "id": "__start",
    "dependencies": []
  },
  {
    "id": "foo1#build",
    "dependencies": [
      "__start",
      "foo2#build"
    ]
  },
  {
    "id": "foo2#build",
    "dependencies": [
      "__start"
    ]
  },
  {
    "id": "foo1#test",
    "dependencies": [
      "__start"
    ]
  },
  {
    "id": "foo2#test",
    "dependencies": [
      "__start",
      "foo2#build"
    ]
  }
]"
`;
